src/lib/string_pool.rs

```diff
@@ -0,0 +1,509 @@
+        let mut vec = BumpVec::new_in(&pool.bump);
+        pool.current_bump_vec.replace(vec).into_bump_slice_mut()
+    }
+
+    /// Resets the current bump vec to the beginning
+    pub(crate) fn clear_current(&self) {
+        self.inner().rent(|v| v.borrow_mut().clear())
+    }
+
+    /// Obtains the length of the current BumpVec.
+    pub(crate) fn len(&self) -> usize {
+        self.inner().rent(|vec| vec.borrow().len())
+    }
+
+    /// Call callback with an immutable buffer of the current BumpVec. This must
+    /// be a callback to ensure that we don't (safely) borrow the slice for
+    /// longer than it stays vaild.
+    pub(crate) fn current_slice<F, R>(&self, mut callback: F) -> R
+        where F: FnMut(&[XML_Char]) -> R
+    {
+        self.inner().rent(|v| callback(v.borrow().as_slice()))
+    }
+
+    /// Call callback with a mutable buffer of the current BumpVec. This must
+    /// be a callback to ensure that we don't (safely) borrow the slice for
+    /// longer than it stays vaild.
+    pub(crate) fn current_mut_slice<F, R>(&self, mut callback: F) -> R
+        where F: FnMut(&mut [XML_Char]) -> R
+    {
+        self.inner().rent(|v| callback(v.borrow_mut().as_mut_slice()))
+    }
+
+    /// Unsafe temporary version of `current_slice()`. This needs to be removed
+    /// when callers are made safe.
+    pub(crate) unsafe fn current_start(&self) -> *const XML_Char {
+        self.inner().rent(|v| v.borrow().as_ptr())
+    }
+
+    /// Appends a char to the current BumpVec.
+    pub(crate) fn append_char(&self, c: XML_Char) -> bool {
+        self.inner().rent(|vec| self.rented_append_char(c, &mut *vec.borrow_mut()))
+    }
+
+    fn rented_append_char(&self, c: XML_Char, vec: &mut BumpVec<XML_Char>) -> bool {
+        if self.rented_is_full(vec) && !self.rented_grow(vec) {
+            false
+        } else {
+            vec.push(c);
+
+            true
+        }
+    }
+
+    /// Overwrites the last char in the current BumpVec.
+    /// Note that this will panic if empty. This is not an insert
+    /// operation as it does not shift bytes afterwards.
+    pub(crate) fn replace_last_char(&self, c: XML_Char) {
+        self.inner().rent(|buf| {
+            *buf.borrow_mut()
+                .last_mut()
+                .expect("Called replace_last_char() when string was empty") = c;
+        })
+    }
+
+    /// Decrements the length, panicing if len is 0
+    pub(crate) fn backtrack(&self) {
+        self.inner().rent(|vec| vec.borrow_mut().pop().expect("Called backtrack() on empty BumpVec"));
+    }
+
+    /// Gets the last character, panicing if len is 0
+    pub(crate) fn get_last_char(&self) -> XML_Char {
+        self.inner().rent(|buf| *buf.borrow().last().expect("Called get_last_char() when string was empty"))
+    }
+
+    /// Appends an entire C String to the current BumpVec.
+    pub(crate) unsafe fn append_c_string(&self, mut s: *const XML_Char) -> bool {
```

Comment:
See discussing for `CStringIterator` below, with the only difference being that `copy_c_string` appends the null terminator while `append_c_string` doesn't.

Reply:
Same thing with my other comment; I'm not sure if we can use `CStringIterator` here because we need to check allocation size first, unless we iterate twice, once to determine length, and once for the extension.